src/v/cluster/cloud_storage_size_reducer.h

```diff
@@ -0,0 +1,196 @@
+ * The request will contain the partitions being queried by shard.
+ * 3. Reduce the reponses and update the accumulator.
+ *
+ * One important thing to note is that this operation requires topic table
+ * stability (i.e. the topic table does not change in the meantime). This
+ * property is enfored by the `topic_table_partition_generator`, which will
+ * throw if the topic table has been updated during the iteration. If such
+ * a change does happen, it's treated as a retryable error.
+ */
+class cloud_storage_size_reducer {
+public:
+    static constexpr uint8_t default_retries_allowed = 3;
+
+    cloud_storage_size_reducer(
+      ss::sharded<topic_table>& topic_table,
+      ss::sharded<members_table>& members_table,
+      ss::sharded<rpc::connection_cache>& conn_cache,
+      size_t batch_size = topic_table_partition_generator::default_batch_size,
+      uint8_t retries_allowed = 3)
+      : _topic_table(std::ref(topic_table))
+      , _members_table(std::ref(members_table))
+      , _conn_cache{conn_cache}
+      , _batch_size{batch_size}
+      , _retries_allowed{retries_allowed} {}
+
+    ss::future<std::optional<uint64_t>> reduce() {
+        try {
+            co_return co_await do_reduce();
+        } catch (const topic_table_partition_generator_exception& e) {
+            vlog(
+              clusterlog.warn,
+              "Failed to generate total cloud storage usage across the cluster "
+              "(retries left: {}): {}",
+              _retries_allowed - _retries_attempted,
+              e.what());
+        } catch (const cloud_storage_size_reducer_exception& e) {
+            vlog(
+              clusterlog.warn,
+              "Failed to generate total cloud storage usage across the cluster "
+              "(retries left: {}): {}",
+              _retries_allowed - _retries_attempted,
+              e.what());
+        } catch (const std::exception& e) {
+            vlog(clusterlog.warn, "Unexpected exception thrown: {}", e.what());
+            throw;
+        }
+
+        if (++_retries_attempted <= _retries_allowed) {
+            co_return co_await reduce();
+        } else {
+            co_return std::nullopt;
+        }
+    }
+
+private:
+    ss::future<uint64_t> do_reduce() {
+        topic_table_partition_generator partition_gen{
+          _topic_table, _batch_size};
+
+        uint64_t total_cloud_storage_bytes{0};
+        auto batch = co_await partition_gen.next_batch();
+        while (batch) {
+            requests_t request_futures = map_batch(*batch);
+            auto results = co_await ss::when_all_succeed(
+              request_futures.begin(), request_futures.end());
+
+            for (auto& res : results) {
+                if (res.has_error()) {
+                    throw cloud_storage_size_reducer_exception(
+                      "RPC for cloud storage usage failed");
+                }
+
+                if (res.value().missing_partitions.size() > 0) {
```

Comment:
we don't actually use the value of `missing_partitions` we just check if its > 0. I guess because then theres a new "generation" of the topics table and we want to start again. Maybe `missing_partitions` can be an integer? or could we actually continue the loop adding those ntps to the list of ntps we need to lookup?

Reply:
> I guess because then theres a new "generation" of the topics table and we want to start again.

`missing_partitions` will be populated in two cases:
* the node has a different topic table generation from the node that issued the request
* the requester and replier have the same topic table generation, but the replier has not yet
applied the changes to the partition manger (see [this comment](https://github.com/redpanda-data/redpanda/pull/9305#discussion_r1129270950) for more details)

> Maybe missing_partitions can be an integer? or could we actually continue the loop adding those ntps to the list of ntps we need to lookup?

I made the response return the list of missing partitions, because it allows for an optimisation. We can retry only the missing partitions instead of the whole request (provided the topic table revision does not change). This is not currently
implemented as I was trying to keep things simple for the first iteration.

Reply:
Added a comment.